File: app/src/test/java/com/example/carlauncher/someip/VSomeIpDataSourceTest.java
package com.example.carlauncher.someip;

import com.example.carlauncher.data.DataSourceStatus;
import com.example.carlauncher.data.DataStatus;
import com.example.carlauncher.data.VehicleDataSource;
import com.example.carlauncher.model.DataValidity;
import com.example.carlauncher.model.Gear;
import com.example.carlauncher.model.TurnSignal;
import com.example.carlauncher.model.VehicleState;
import com.example.carlauncher.model.WarningState;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

public class VSomeIpDataSourceTest {

    private FakeNativeVehicleTransport fakeTransport;
    private SomeipConnectionMonitor connectionMonitor;
    private VSomeIpDataSource dataSource;
    private TestVehicleDataSourceListener listener;

    @Before
    public void setUp() {
        fakeTransport = new FakeNativeVehicleTransport();
        connectionMonitor = new SomeipConnectionMonitor();
        dataSource = new VSomeIpDataSource(fakeTransport, connectionMonitor);
        listener = new TestVehicleDataSourceListener();
    }

    @Test
    public void start_subscribesToTransportAndNotifiesConnecting() {
        dataSource.start(listener);

        assertTrue(dataSource.isRunning());
        assertEquals(DataSourceStatus.CONNECTING, dataSource.getStatus());
        assertTrue(listener.sourceStatuses.contains(DataSourceStatus.CONNECTING));
    }

    @Test
    public void availabilityChanged_updatesStatusAndMonitor() {
        dataSource.start(listener);

        fakeTransport.notifyAvailability(true);
        assertEquals(DataSourceStatus.CONNECTING, dataSource.getStatus());

        fakeTransport.notifyAvailability(false);
        assertEquals(DataSourceStatus.DISCONNECTED, dataSource.getStatus());
        assertNull(dataSource.getLatestState());
    }

    @Test
    public void vehicleState_updatesLatestStateAndStatusConnected() {
        dataSource.start(listener);
        fakeTransport.notifyAvailability(true);

        VehicleState sampleState = createSampleVehicleState(1, 1000L, 60, DataStatus.NORMAL);
        fakeTransport.notifyVehicleState(sampleState);

        assertEquals(sampleState, dataSource.getLatestState());
        assertEquals(DataSourceStatus.CONNECTED, dataSource.getStatus());
        assertEquals(1, listener.receivedStates.size());
        assertEquals(sampleState, listener.receivedStates.get(0));
    }

    @Test
    public void vehicleState_withNoData_updatesStatusNoData() {
        dataSource.start(listener);
        fakeTransport.notifyAvailability(true);

        VehicleState sampleState = createSampleVehicleState(2, 2000L, 0, DataStatus.NO_DATA);
        fakeTransport.notifyVehicleState(sampleState);

        assertEquals(DataSourceStatus.NO_DATA, dataSource.getStatus());
    }

    @Test
    public void error_updatesStatusErrorAndNotifiesListener() {
        dataSource.start(listener);

        Exception testError = new RuntimeException("Test protocol error");
        fakeTransport.notifyError(testError);

        assertEquals(DataSourceStatus.ERROR, dataSource.getStatus());
        assertEquals(1, listener.receivedErrors.size());
        assertEquals(testError, listener.receivedErrors.get(0));
    }

    @Test
    public void stop_unsubscribesAndResetsStatus() {
        dataSource.start(listener);
        fakeTransport.notifyAvailability(true);

        dataSource.stop();

        assertFalse(dataSource.isRunning());
        assertEquals(DataSourceStatus.STOPPED, dataSource.getStatus());
        assertNull(dataSource.getLatestState());
    }

    // ---- Helper Classes ----

    private static VehicleState createSampleVehicleState(long seq, long timestamp, int speed, DataStatus status) {
        return new VehicleState.Builder()
                .setVersion(1)
                .setSequence(seq)
                .setTimestampMs(timestamp)
                .setVehSpeedKph(speed)
                .setEngRpm(2000)
                .setSoc(80)
                .setGear(Gear.D)
                .setTurnSignal(TurnSignal.NONE)
                .setWarning(WarningState.NONE)
                .setValidity(DataValidity.VALID)
                .setDataStatus(status)
                .setParkingBrake(false)
                .setDoorLock(true)
                .setBeltWarning(false)
                .setHeadlightsState(0)
                .setHighBeamLightsState(0)
                .setEngineCoolantTemp(90.0f)
                .setEvBatteryLevel(80.0f)
                .build();
    }

    private static class FakeNativeVehicleTransport implements NativeVehicleTransport {
        private Listener listener;
        private boolean available = false;
        private boolean started = false;

        @Override
        public boolean start() {
            started = true;
            return true;
        }

        @Override
        public void stop() {
            started = false;
            available = false;
        }

        @Override
        public boolean isAvailable() {
            return available;
        }

        @Override
        public void setListener(Listener listener) {
            this.listener = listener;
        }

        void notifyAvailability(boolean available) {
            this.available = available;
            if (listener != null) {
                listener.onAvailabilityChanged(available);
            }
        }

        void notifyVehicleState(VehicleState state) {
            if (listener != null) {
                listener.onVehicleState(state);
            }
        }

        void notifyError(Throwable throwable) {
            if (listener != null) {
                listener.onError(throwable);
            }
        }
    }

    private static class TestVehicleDataSourceListener implements VehicleDataSource.Listener {
        final List<VehicleState> receivedStates = new ArrayList<>();
        final List<DataSourceStatus> sourceStatuses = new ArrayList<>();
        final List<Exception> receivedErrors = new ArrayList<>();

        @Override
        public void onStateChanged(VehicleState state) {
            receivedStates.add(state);
        }

        @Override
        public void onSourceStatusChanged(DataSourceStatus status) {
            sourceStatuses.add(status);
        }

        @Override
        public void onError(Exception exception) {
            receivedErrors.add(exception);
        }
    }
}

File: app/src/test/java/com/example/carlauncher/someip/VSomeIpDataSourceTest.kt
package com.example.carlauncher.someip

import com.example.carlauncher.data.DataSourceStatus
import com.example.carlauncher.data.DataStatus
import com.example.carlauncher.data.VehicleDataSource
import com.example.carlauncher.model.DataValidity
import com.example.carlauncher.model.Gear
import com.example.carlauncher.model.TurnSignal
import com.example.carlauncher.model.VehicleState
import com.example.carlauncher.model.WarningState
import java.util.ArrayList
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

open class VSomeIpDataSourceTest {

    private var fakeTransport: FakeNativeVehicleTransport? = null
    private var connectionMonitor: SomeipConnectionMonitor? = null
    private var dataSource: VSomeIpDataSource? = null
    private var listener: TestVehicleDataSourceListener? = null

    @Before
    open fun setUp() {
        fakeTransport = FakeNativeVehicleTransport()
        connectionMonitor = SomeipConnectionMonitor()
        dataSource = VSomeIpDataSource(fakeTransport, connectionMonitor)
        listener = TestVehicleDataSourceListener()
    }

    @Test
    open fun start_subscribesToTransportAndNotifiesConnecting() {
        dataSource!!.start(listener)

        assertTrue(dataSource!!.isRunning)
        assertEquals(DataSourceStatus.CONNECTING, dataSource!!.status)
        assertTrue(listener!!.sourceStatuses.contains(DataSourceStatus.CONNECTING))
    }

    @Test
    open fun availabilityChanged_updatesStatusAndMonitor() {
        dataSource!!.start(listener)

        fakeTransport!!.notifyAvailability(true)
        assertEquals(DataSourceStatus.CONNECTING, dataSource!!.status)

        fakeTransport!!.notifyAvailability(false)
        assertEquals(DataSourceStatus.DISCONNECTED, dataSource!!.status)
        assertNull(dataSource!!.latestState)
    }

    @Test
    open fun vehicleState_updatesLatestStateAndStatusConnected() {
        dataSource!!.start(listener)
        fakeTransport!!.notifyAvailability(true)

        val sampleState = createSampleVehicleState(1, 1000L, 60, DataStatus.NORMAL)
        fakeTransport!!.notifyVehicleState(sampleState)

        assertEquals(sampleState, dataSource!!.latestState)
        assertEquals(DataSourceStatus.CONNECTED, dataSource!!.status)
        assertEquals(1, listener!!.receivedStates.size.toLong())
        assertEquals(sampleState, listener!!.receivedStates.get(0))
    }

    @Test
    open fun vehicleState_withNoData_updatesStatusNoData() {
        dataSource!!.start(listener)
        fakeTransport!!.notifyAvailability(true)

        val sampleState = createSampleVehicleState(2, 2000L, 0, DataStatus.NO_DATA)
        fakeTransport!!.notifyVehicleState(sampleState)

        assertEquals(DataSourceStatus.NO_DATA, dataSource!!.status)
    }

    @Test
    open fun error_updatesStatusErrorAndNotifiesListener() {
        dataSource!!.start(listener)

        val testError = RuntimeException("Test protocol error")
        fakeTransport!!.notifyError(testError)

        assertEquals(DataSourceStatus.ERROR, dataSource!!.status)
        assertEquals(1, listener!!.receivedErrors.size.toLong())
        assertEquals(testError, listener!!.receivedErrors.get(0))
    }

    @Test
    open fun stop_unsubscribesAndResetsStatus() {
        dataSource!!.start(listener)
        fakeTransport!!.notifyAvailability(true)

        dataSource!!.stop()

        assertFalse(dataSource!!.isRunning)
        assertEquals(DataSourceStatus.STOPPED, dataSource!!.status)
        assertNull(dataSource!!.latestState)
    }

    // ---- Helper Classes ----

    private fun createSampleVehicleState(
        seq: Long,
        timestamp: Long,
        speed: Int,
        status: DataStatus?,
    ): VehicleState? =
        VehicleState.Builder()
            .setVersion(1)
            .setSequence(seq)
            .setTimestampMs(timestamp)
            .setVehSpeedKph(speed)
            .setEngRpm(2000)
            .setSoc(80)
            .setGear(Gear.D)
            .setTurnSignal(TurnSignal.NONE)
            .setWarning(WarningState.NONE)
            .setValidity(DataValidity.VALID)
            .setDataStatus(status)
            .setParkingBrake(false)
            .setDoorLock(true)
            .setBeltWarning(false)
            .setHeadlightsState(0)
            .setHighBeamLightsState(0)
            .setEngineCoolantTemp(90.0f)
            .setEvBatteryLevel(80.0f)
            .build()

    private open class FakeNativeVehicleTransport : NativeVehicleTransport {
        private var listener: NativeVehicleTransport.Listener? = null
        final override var isAvailable: Boolean = false
            private set

        private var started: Boolean = false

        override fun start(): Boolean {
            started = true
            return true
        }

        override fun stop() {
            started = false
            isAvailable = false
        }

        override fun setListener(listener: NativeVehicleTransport.Listener?) {
            this.listener = listener
        }

        internal open fun notifyAvailability(available: Boolean) {
            this.isAvailable = available
            if (listener != null) {
                listener!!.onAvailabilityChanged(available)
            }
        }

        internal open fun notifyVehicleState(state: VehicleState?) {
            if (listener != null) {
                listener!!.onVehicleState(state)
            }
        }

        internal open fun notifyError(throwable: Throwable?) {
            if (listener != null) {
                listener!!.onError(throwable)
            }
        }
    }

    private open class TestVehicleDataSourceListener : VehicleDataSource.Listener {
        internal val receivedStates: MutableList<VehicleState?> = ArrayList<VehicleState?>()
        internal val sourceStatuses: MutableList<DataSourceStatus?> = ArrayList<DataSourceStatus?>()
        internal val receivedErrors: MutableList<Exception?> = ArrayList<Exception?>()

        override fun onStateChanged(state: VehicleState?) {
            receivedStates.add(state)
        }

        override fun onSourceStatusChanged(status: DataSourceStatus?) {
            sourceStatuses.add(status)
        }

        override fun onError(exception: Exception?) {
            receivedErrors.add(exception)
        }
    }
}
